Allow Sherpa to build against XSPEC 12.15.0

DougBurke opened the pull request:
# Summary

Allow building against XSPEC 12.15.0. Adds the following 17 models:

- additive: XSfeklor, XSvlorentz, XSvvoigt, XSzfeklor, XSzlorentz, XSzvlorentz, XSzvoigt, XSzvvoigt
- convolution: XSrgsext
- multiplicative: XSlorabs, XSvlorabs, XSvoigtabs, XSvvoigtabs, XSzlorabs, XSzvorlabs, XSzvoigtabs, XSzvvoigtabs

The XSPEC initialization file (~/.xspec/Xspec.init) may need to be removed or the AtomDB version in it updated before using XSPEC models.

# Details

The first commit updates the scripts `add_xspec_model.py` and `update_xspec_functions.py` to reduce friction. The `sherpa.astro.utils.xspec` code used by `add_xspec_model.py` is primarily intended to support XSPEC user models, and so the output required manipulation when used for XSPEC, so the code has been updated to support this "internal" use to better match the Python code used in `sherpa.astro.xspec` (adding a version decorator, a "versionchanged" note, and a reference link with a best-guess for the XSPEC HTML documentation location). The `update_xspec_functions.py` code preciously did not recognize version constraints - the `#ifdef XSPEC_x_y_z` lines used to indicate function changes over time for a model - because there's no encoding of this information in XSPEC itself. The code now takes the existing `sherpa/astro/xspec/src/_xspec.cc` file and uses that to define the version constraints for each model, and uses that to create the new output. It does not create the smallest possible set of `#ifdef` lines, but it is a huge improvement over the previous approach. 

The second commit adds a new helper script - `scripts/update_xspec_docs.py`` which automates the update of the ``docs/model_classes/astro_xspec.rst`` file. It turns out I missed some changes in the XSPEC 12.14.x series so this also updates the ``astro_xspec.rst`` file to reflect these new models.

The last commit adds support for XSPEC 12.15.0. Unfortunately you can not build against 12.15.0 without these changes. There are new models, but no parameter changes.

When testing

- you may need to update or remove `~/.xspec/Xspec.init` to addres AtomDb version updates
- the tests force a particular AtomDB version (3.0.9) that was used in recent XSPEC releases (the default for 12.15.0 is 3.1.2 or so); this is not ideal and we need something like #2224 to handle this more gracefully

There's no conda build of 12.15.0 so we can not test this with the CI runs yet. 

# Notes

The "biggest issue" I see here is the "let's just run all XSPEC tests with AtomDB 3.0.9" pytest fixture that is added. I see this as a stop-gap until either - or both - of 

- XSPEC 12.15.0 is the oldest version we expect to build against
- we get something like #2224 merged which gives direct access to the AtomDB version used by XSPEC and we can relpace a global fixture with a more test-appropriate one (this cuold technically be done here but it's more awkward to do)

DougBurke commented:
Rebased as #2249 has been merged and to fix a few minor issues in the model docstrings.

hamogu commented on docs/developer/index.rst:
```diff
@@ -394,18 +392,21 @@ version then you can use three helper scripts:
-   for the class.
+   Python (``sherpa/astro/xspec/__init__.py``). The Python code
+   lacks documentation and some values either need adding (e.g.
+   the Sherpa version) or lnks checked and possibly updated (due
```
links -> links

hamogu commented on scripts/update_xspec_functions.py:
```diff
@@ -36,68 +36,450 @@
-        return prev
 
-    # For the moment models that are calculated per-spectrum are unsupported.
+def convert_version(smajor: str, sminor: str, smicro: str) -> Version:
+    """Convert version tuple"""
+
```
Isn't this identical to the function in `add_xspec_model.py`? If so, write once and import. But maybe that's changed in a latter commit; hard to know when reviewing commit-by-commit. Might have other get back to this comment again in the end.

DougBurke replied:
So, I did think about moving some code into a "helper" module that these scripts could import, but I decided that I didn't want to spend the extra time to do it.

hamogu commented on scripts/update_xspec_functions.py:
```diff
@@ -36,68 +36,450 @@
+
+    # Can we find it from the installed location (which implies an
+    # editable install)? This is not ideal, since it is somewhat
+    # mis-using the importlib machinery. The files is labelled as
```
The files is labeled -> The file (singular)

DougBurke replied:
Actually, it should be "The files call is labelled as" as I'm meant to be talking about `resources.files` here. An amendment will be made.

hamogu reviewed (commented):
This is really hard to review for me because I've never build XSPEC myself and I've never build sherpa against anything but the xspec models as provided by DS as conda packages. For a real review, I will have to test locally against at least two versions of xspec.
In particular since I don't fully understand (and admittedly have so far taken any effort to not spend time on learning this) how our XSPEC model wrappers actually work - unfortunately I can't say from just reading the Python code that it's all good because I don't actually know how "all good" would look like.

So, the question is how to move forward. Do I actually need to fully understand all the helper scripts or can we agree that "the proof is in the pudding" in the sense that (a) the scripts say "review output manually" and (b) @DougBurke used them to make a working version of XSPEC 12.15, so apparently they do mostly (I understand that there are still some areas that might need manual editing of the output) work?

Similarly, do we think that a manual review of the newly added XSPEC model wrappers is actually useful or is it sufficient that they actually pass our tests? We usually don't test the numerics of those models (if only, because we don't have a ground truth for many of them, that's XSPEC's job, not ours).

In other words, how much scrutiny is actually required to mark a PR like this as "approved"?

DougBurke commented:
So, the aim is to try and make this as "mechanical" a process as possible, to remove the need to worry about chnages (as long as the tests keep passing). Of course, without an XSPEC 12.15.0 conda build our CI tests only check we can build against older versions (which is useful, but not complete).

I had meant to ask @nplee to see if he could build against XSPEC 12.15.0 and run tests (in particular on a macOS arm machine if possible, since that is where I'd expect to possibly see differences). So hopefully Nick can provide some checks.

What I am trying to get to is something like #52 - but there are reasons why it's not easy from both the XSPEC side (we can automate some of creating `_xspec.cc` but I'm not as sure about the Python code), and from the having-to-fight-with-setuptools side (I'm hopeful that meson provides beter traceability to allow generated code to be used). This is all somewhat irrelevant to this PR, but I may try something based on https://github.com/sherpa/sherpa/issues/52#issuecomment-2761251855 - whci is also irrelevant to this.

I had tried to help the review out by pointing out areas I was interested in talking about, but I obviously need to be more direct.

hamogu commented:
> I had tried to help the review out by pointing out areas I was interested in talking about, but I obviously need to be more direct.

No, it's fine, I will do it and let you know if I need help. But, for the reasons stated, it's not going to happen today.

DougBurke commented:
@nplee - I thought I'd rebased this to pick up all the "setuptools 78" nastiness for this PR, but maybe I missed #2255. 

For the DS9 stuff, throw up an issue once you've got some idea. I do know that I stopped DS9 installations for our CI runs on macOS, but I had assumed that was some issue with how they set up the "virtual" environment. Maybe there's a larger issue with macOS, X, and the like?

hamogu commented:
There are really two issues wrapped in one in this PR (because they are intertwined):
- improve the scripts that generate part of the code
- include the output (with some hand-editing) of those scripts into the codebase to be able to build against XSPEC 12.15.0.

@nplee confirms that the latter part works - and the fact that it does also shows that the first part is OK or at least not harmful - the scripts don't have to be perfect as long as we expect to hand-check their outputs to include them.

For this PR, we should take it like that, but I have some more general thoughts on the structure of the scripts that might be interested in the context of #52, so I'll comment there.

nplee commented:
Just to add that if I don't update the ATOMDB/NEI version in the Xspec.init file when I'm using XSPEC 12.15.0 I see that 80 of the tests fail.

nplee commented:
@DougBurke The difference between the 12.14.1 and 12.15.0 `Xspec.init` file that Sherpa picks up at XSpec initialization is

```
< ATOMDB_VERSION:  3.0.9
< NEI_VERSION: 3.0.4
---
> ATOMDB_VERSION:  3.1.2
> SPEX_VERSION:  3.08
> NEI_VERSION: 3.1.2
```

We can modify the tests:

`sherpa/astro/xspec/tests/test_xspec.py::test_evaluate_xspec_model`
and
`sherpa/astro/xspec/tests/test_xspec.py::test_evaluate_xspec_model_noncontiguous2`

by defining the AtomDB version being user in `APECROOT` and `NEIVERS`...

```python
### XSPEC 12.15.0 transitions to using ATOMDB 3.1.2 while 12.14 were dependent on 3.0.9
### and the NEI version remained static between 3.0.4 and 3.0.9
def set_atomdbver():
    """ set APEC/NEI ATOMDB versioning information """
    from sherpa.astro.xspec import get_xsversion, set_xsxset
    
    xsver = get_xsversion()

    if xsver >= "12.15.0":
        set_xsxset("APECROOT","3.1.2")
        set_xsxset("NEIVERS", "3.1.2")
    else:
        set_xsxset("APECROOT","3.0.9")
        set_xsxset("NEIVERS", "3.0.4")

...

@requires_xspec
@pytest.mark.parametrize("modelcls", get_xspec_models())
def test_evaluate_xspec_model(modelcls):
    """Can we call a model with its default parameters?

    There is limited validation of the results. A regression-style test could
    be made (e.g. with use of pytest-arraydiff) but is it worth it?

    Rather than loop over each model within the test, use pytest to loop over
    each model. This means that make_grid gets called multiple times, but
    (hopefully) makes it a lot easier to debug when only a subset of tests
    fails.  It also makes sure that all models are run, even if some fail,
    which did not happen when they were all stuck in the same test function.

    Convolution models are skipped (easier to filter out here given the
    current design).
    """

    from sherpa.astro import xspec

    # use an identifier in case there is an error
    mdl = modelcls()
    if isinstance(mdl, xspec.XSConvolutionKernel):
        return

    # set AtomDB versioning
    set_atomdbver()
    
    elo, ehi, wlo, whi = make_grid()

    # The model checks that the values are all finite,
    # so there is no need to check that the output of
    # mdl does not contain non-finite values.
    #
    evals = mdl(elo, ehi)
    wvals = mdl(wlo, whi)

    assert_is_finite(evals, modelcls, "energy")
    assert_is_finite(wvals, modelcls, "wavelength")

    assert wvals == pytest.approx(evals)


@requires_xspec
@pytest.mark.parametrize("modelcls", get_xspec_models())
def test_evaluate_xspec_model_noncontiguous2(modelcls):
    """Can we evaluate an XSPEC model with a non-contiguous grid?

    Note that there is no test that the non-contiguous grid
    results are similar to the result from using a contiguous
    grid and then filtering out the missing bins. The way that
    some models are implemented make this a tricky test to
    write (due to numerical tolerances), as bins at the
    edges may not match well.
    """

    from sherpa.astro import xspec

    # use an identifier in case there is an error
    mdl = modelcls()
    if isinstance(mdl, xspec.XSConvolutionKernel):
        return
    
    # set AtomDB versioning
    set_atomdbver()

    elo, ehi, wlo, whi = make_grid_noncontig2()

    evals2 = mdl(elo, ehi)
    wvals2 = mdl(wlo, whi)

    assert_is_finite(evals2, modelcls, "energy")
    assert_is_finite(wvals2, modelcls, "wavelength")

    assert wvals2 == pytest.approx(evals2)

```
and allow us to move forward.  Albeit, maybe it's a good thing the tests fail and forces the user to be aware of the XSpec/AtomDB versioning discrepancies?

DougBurke commented:
@nplee - this PR comes with a change to `sherpa/conftest.py` which does the APECROOT change but not the NEIVERS one. Can you try updating it to see if adding

```python
xspec.set_xsxset("NEIVERS", "3.0.4")
```

works without your changes (I expect a number of serialization tests to fail). It maybe we only need to apply this to the `test_xspec.py` file but I want to check things out.

nplee commented:
@DougBurke you're correct that this causes some of the serialization tests to fail, and I had to do some experimentation using a build with XSpec 12.14.1k and the `Xspec.init` file for 12.15.0 and another build using 12.15.0 while using the earlier initialization file, but it does look like `sherpa/conftest.py` applies the AtomDB settings to all the other tests.  We have to also define `NEIAPECROOT` to point to the APEC NEI file version since `NEIVERS` applies to the version of the NEI eigenvector file.

```python
@pytest.fixture(scope="session", autouse=True)
def set_xspec_atomdb_version():
    """Ensure the AtomDB abundance is 3.0.9 if XSPEC is enabled."""

    if not has_xspec:
        return

    # XSPEC 12.15.0 switches the AtomDB version from 3.0.9 to 3.1.2
    # which can changes enough to cause the test to fail. The
    # tolerances could be changed but instead try setting the AtomDB
    # version. Unfortunately the current interface is rather limited
    # (see #2216), so this is done globally.
    #
    if xspec.get_xsversion() >= "12.15.0":
        xspec.set_xsxset("APECROOT", "3.1.2") 
        xspec.set_xsxset("NEIAPECROOT", "3.1.2") 
        xspec.set_xsxset("NEIVERS", "3.1.2") 
    else:
        xspec.set_xsxset("APECROOT", "3.0.9") # equilbrium 'apec_v<version>'
        xspec.set_xsxset("NEIAPECROOT", "3.0.9") # non-equilibrium 'apec_v<version>_nei'
        xspec.set_xsxset("NEIVERS", "3.0.4") # non-equilibrium eigenvector file version
```

The XSpec tests pass with this modification for the two builds mentioned above.

nplee commented:
@DougBurke that is to say the APEC/NEI dependent tests in: 

`sherpa/astro/xspec/tests/test_xspec.py::test_evaluate_xspec_model` 
and 
`sherpa/astro/xspec/tests/test_xspec.py::test_evaluate_xspec_model_noncontiguous2`

pass but with some failures in `sherpa/astro/ui/tests/test_serialize.py`, as expected, that have assertions errors with the APEC versioning being specified, unless I'm missing something in what you're saying...


```python
(sherpa-4.17-standalone-py311-xs12.15.0) falafel-296: pytest $CONDA_PREFIX/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py 
======================================= test session starts ========================================
platform linux -- Python 3.11.11, pytest-8.3.3, pluggy-1.5.0
rootdir: /scratch2
plugins: anyio-4.9.0
collected 48 items                                                                                 

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py . [  2%]
.......F.F.F........F......FF....x....F........                                              [100%]

============================================= FAILURES =============================================
_____________________________________ test_canonical_pha_basic _____________________________________

make_data_path = <function make_data_path.<locals>.wrapped at 0x7f5e981db7e0>

    @requires_data
    @requires_xspec
    @requires_fits
    def test_canonical_pha_basic(make_data_path):
    
        _, canonical = setup_pha_basic(make_data_path)
>       compare(canonical)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2646: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha(1, "/scratch2/Anaconda/anaconda3/...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha(1, "/scratch2/Anaconda/anaconda3/...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
____________________________________ test_canonical_pha_grouped ____________________________________

make_data_path = <function make_data_path.<locals>.wrapped at 0x7f5e981db7e0>

    @requires_data
    @requires_xspec
    @requires_fits
    @requires_group
    def test_canonical_pha_grouped(make_data_path):
    
        _, _, canonical = setup_pha_grouped(make_data_path)
>       compare(canonical)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2680: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("grp", "/scratch2/Anaconda/anacon...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("grp", "/scratch2/Anaconda/anacon...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
_____________________________________ test_canonical_pha_back ______________________________________

make_data_path = <function make_data_path.<locals>.wrapped at 0x7f5e981db7e0>

    @requires_data
    @requires_xspec
    @requires_fits
    @requires_group
    def test_canonical_pha_back(make_data_path):
    
        _, _, canonical = setup_pha_back(make_data_path)
>       compare(canonical)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2723: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("bgrp", "/scratch2/Anaconda/anaco...xschatter(0)\nset_xsabund("lodd")\nset_xscosmo(72, 0.02, 0.71)\nset_xsxsect("vern")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("bgrp", "/scratch2/Anaconda/anaco...xsxsect("vern")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
--------------------------------------- Captured stdout call ---------------------------------------
 Solar Abundance Vector set to lodd:  Lodders, K. ApJ 591, 1220 (2003)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
____________________________________ test_restore_xstable_model ____________________________________

make_data_path = <function make_data_path.<locals>.wrapped at 0x7f5e981db7e0>

    @requires_xspec
    @requires_data
    @requires_fits
    def test_restore_xstable_model(make_data_path):
        """Note: this only sets the table model
    
        This is a regression test as it currently does not do the right
        thing.
    
        """
    
        ui.load_xstable_model("tbl", make_data_path('testpcfabs.mod'))
        tbl.nh = 2
        tbl.fract = 0.2
        ui.freeze(tbl.fract)
    
        # Comparing the string version is a simlple way to check the
        # parameters.
        #
        orig = str(tbl)
    
>       compare(add_datadir_path(_canonical_xstable_model))

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:3027: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
_______________________________ test_canonical_xspec_hard_limit_min ________________________________

    @requires_xspec
    def test_canonical_xspec_hard_limit_min():
        "Can we save an XSPEC model with the hard limit extended: min"
    
        # Reset the optimiser parameters to make them easy to check,
        # even if they are un-usable.
        #
        for key in ui.get_method_opt().keys():
            ui.set_method_opt(key, 1)
    
        ui.create_model_component('xspowerlaw', 'mdl')
        mdl.phoindex.set(val=-5, hard_min=-5, frozen=True)
        mdl.norm.max = 100
    
>       compare(_canonical_xspec_hard_limit_min)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:3189: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
_______________________________ test_canonical_xspec_hard_limit_max ________________________________

    @requires_xspec
    def test_canonical_xspec_hard_limit_max():
        "Can we save an XSPEC model with the hard limit extended: max"
    
        # Reset the optimiser parameters to make them easy to check,
        # even if they are un-usable.
        #
        for key in ui.get_method_opt().keys():
            ui.set_method_opt(key, 1)
    
        ui.create_model_component('xspowerlaw', 'mdl')
        mdl.phoindex.set(val=15, hard_max=15, frozen=True)
        mdl.norm.max = 100
    
>       compare(_canonical_xspec_hard_limit_max)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:3206: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n\n######### Set Statistic\n\nset_stat("chi2gehrels")\n\n\n######### Se...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
_______________________________________ test_restore_pha_csc _______________________________________

make_data_path = <function make_data_path.<locals>.wrapped at 0x7f5e981db7e0>

    @requires_data
    @requires_group
    @requires_fits
    @requires_xspec
    def test_restore_pha_csc(make_data_path):
        """Can we restore a CSC file.
    
        This just checks that results of reading in. It doesn't
        check the actual text of the saved file.
        """
    
        ui.set_stat("cstat")
        ui.set_method("simplex")
        ui.load_pha("csc", make_data_path("acisf01575_001N001_r0085_pha3.fits.gz"))
    
        # We can check out some specialized cases, where we have a
        # different filter for the source and background data. Even though
        # we are using cstat we still bin (to 3 counts per bin).
        #
        # Note that this data is ungrouped.
        #
        # We check the sum of the grouping array as a reasonable way to
        # check the grouping is correct, without having to check all 1024
        # elements.
        #
        expected_sgrp = -110
        expected_bgrp = -457
    
        ui.notice(0.5, 7)
        ui.group_counts("csc", 3, tabStops=~ui.get_data("csc").get_mask())
        assert ui.get_grouping("csc").sum() == expected_sgrp
    
        ui.notice_id("csc", 0.4, 8, bkg_id=1)
        ui.group_counts("csc", 3, tabStops=~ui.get_bkg("csc").get_mask(), bkg_id=1)
        assert ui.get_grouping("csc", bkg_id=1).sum() == expected_bgrp
        assert ui.get_grouping("csc").sum() == expected_sgrp  # check it hasn't changed
    
        sfilter = "0.50:7.01"
        bfilter = "0.39:8.00"
        assert ui.get_data("csc").get_filter(format="%.2f") == sfilter
        assert ui.get_bkg("csc").get_filter(format="%.2f") == bfilter
    
        ui.set_bkg_source("csc", ui.powlaw1d.bpl)
        ui.set_source("csc", ui.xsphabs.gal * ui.powlaw1d.spl)
    
        bpl.gamma = 0.87
        bpl.ampl = 2.14e-6
        spl.gamma = 1.64
        spl.ampl = 3.9e-5
        gal.nh = 0.24
    
        expected_stat = 185.3380144554129
        assert ui.calc_stat("csc") == pytest.approx(expected_stat)
    
        expected_output = add_datadir_path(_canonical_pha_csc)
>       compare(expected_output)

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:3630: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2427: in compare
    compare_lines(expected, output)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("csc", "/scratch2/Anaconda/anacon...et_xschatter(0)\nset_xsabund("angr")\nset_xscosmo(70, 0, 0.73)\nset_xsxsect("bcmc")\nset_xsxset("APECROOT", "3.0.9")\n'
got = 'import numpy\nfrom sherpa.astro.ui import *\n\n######### Load Data Sets\n\nload_pha("csc", "/scratch2/Anaconda/anacon...xsxsect("bcmc")\nset_xsxset("APECROOT", "3.1.2")\nset_xsxset("NEIAPECROOT", "3.1.2")\nset_xsxset("NEIVERS", "3.1.2")\n'

    def compare_lines(expected, got):
        """Check that each line in got matches that in
        expected. This is to provide a more-readable
        set of error messages (may prefer a diff-style
        analysis).
        """
    
        elines = expected.split('\n')
        glines = got.split('\n')
    
        for e, g in zip(elines, glines):
>           assert e == g, f"expected='{e}' got='{g}'"
E           AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
E           assert 'set_xsxset("...OT", "3.0.9")' == 'set_xsxset("...OT", "3.1.2")'
E             
E             - set_xsxset("APECROOT", "3.1.2")
E             ?                           ^ ^
E             + set_xsxset("APECROOT", "3.0.9")
E             ?                           ^ ^

../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py:2404: AssertionError
-------------------------------------- Captured stdout setup ---------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to bcmc:  Balucinska-Church and McCammon, 1998
------------------------------------- Captured stdout teardown -------------------------------------
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
 Solar Abundance Vector set to angr:  Anders E. & Grevesse N. Geochimica et Cosmochimica Acta 53, 197 (1989)
 Cross Section Table set to vern:  Verner, Ferland, Korista, and Yakovlev 1996
===================================== short test summary info ======================================
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_canonical_pha_basic - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_canonical_pha_grouped - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_canonical_pha_back - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_restore_xstable_model - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_canonical_xspec_hard_limit_min - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_canonical_xspec_hard_limit_max - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
FAILED ../../Anaconda/anaconda3/envs/sherpa-4.17-standalone-py311-xs12.15.0/lib/python3.11/site-packages/sherpa/astro/ui/tests/test_serialize.py::test_restore_pha_csc - AssertionError: expected='set_xsxset("APECROOT", "3.0.9")' got='set_xsxset("APECROOT", "3.1.2")'
============================= 7 failed, 40 passed, 1 xfailed in 17.09s =============================
```

DougBurke commented:
@nplee - I'm hoping the latest version will work with 12.15.0, but YMMV

DougBurke commented:
@nplee Ta - the current "force an ATOMDB/..." version is not long-term viable, but should work for now. Can you create an issue for your setuptools problem as I don't seem to be seeing this (or our CI runs) so I don't understand it.

hamogu reviewed (approved):
Me approving this here is really just a summary of the work by @nplee who did run this PR on multiple versions and platforms and it passes. That's the important part - that the Sherpa code itself works.

The changes in the helper scripts that can help to prepare Sherpa for XSPEC updates are in comparison less important - we expect them to not work perfectly and thus manually review the changes (which @DougBurke and I did and @nplee confirmed). The changes in this PR are a step towards the more and more automatic generation of more and more of the XSPEC models even if they don't reach that end goal at this point (see #52).

DougBurke commented:
@hamogu - I have resolved conflicts (fortunately easy to do without a merge) and addressed points from @wmclaugh . However, I've taken the time to add two new commits which I first wrote for #2275 since they are related to the scripts clean up I added here (add tests to show a small "problem" - that I didn't add " convolution" to a string because of a stupid check) and then fix that (along with another logical error that isn't actually an error in practice, coupled with some mechanical typing changes). Are you able to review the last two commits or should I pull them?

hamogu reviewed (approved):
I still approve even with the two new commits added (assuming CI passes).

wmclaugh commented on sherpa/astro/utils/tests/test_astro_utils_xspec.py:
```diff
@@ -548,6 +635,83 @@ def __init__(self, name='abcd'):
 
+@requires_xspec
+def test_create_model_multiplicative_xspec(caplog):
+    """Fake up a multplicative model for XSPEC"""
```
```suggestion
    """Fake up a multiplicative model for XSPEC"""
```

wmclaugh commented on sherpa/astro/utils/tests/test_astro_utils_xspec.py:
```diff
@@ -483,6 +484,91 @@ def __init__(self, name='apec'):
+
 @requires_xspec
 def test_create_model_multiplicative(caplog):
     """Fake up a multplicative model"""
```
```suggestion
    """Fake up a multiplicative  model"""
```

wmclaugh commented on sherpa/astro/utils/tests/test_astro_utils_xspec.py:
```diff
@@ -635,6 +635,83 @@ def __init__(self, name='abcd'):
+    References
+    ----------
+
+    .. [1] https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSmodelAbcd.html
```
This URL doesn't seem to work- returns "Sorry, but the document you requested is not available on this server"

DougBurke replied:
Yeah - that's by intent, as 

- this is for a test and it is just checking that the model name has been capitalized correctly (`abcd` -> `Abcd`)
- the output is for text that needs to be checked before it can be directly used, as - thanks to the way that XSPEC models are documented - we can not guarantee that `XSmodel<name>.html` is valid, but it's going to be close to being the truth

DougBurke commented on docs/developer/index.rst:
```diff
@@ -394,18 +392,29 @@ version then you can use three helper scripts:
-   for the class.
+   Python (``sherpa/astro/xspec/__init__.py``). The Python code
+   lacks documentation and some values either need adding (e.g.
+   the Sherpa version) or lnks checked and possibly updated (due
```
```suggestion
   the Sherpa version) or links checked and possibly updated (due
```
